raiden/network/transport/matrix/rtc/web_rtc.py

```diff
@@ -44,21 +46,81 @@ def schedule_task(
         self.coroutines.append(task)
         return task
 
-    async def wait_for_coroutines(self) -> None:
-        log.debug("Waiting for coroutines", coroutines=self.coroutines)
-        await asyncio.gather(*self.coroutines)
+    async def wait_for_coroutines(self, cancel: bool = True) -> None:
+
+        if cancel:
+            self.cancel_all_pending()
+
+        pending_coroutines = [
+            coroutine
+            for coroutine in self.coroutines
+            if not coroutine.done() and not coroutine.cancelled()
+        ]
+        log.debug("Waiting for coroutines", coroutines=pending_coroutines)
+
+        try:
+            await asyncio.gather(*pending_coroutines)
+        except CancelledError:
+            log.debug(
+                "Pending coroutines cancelled",
+                cancelled=[coroutine for coroutine in self.coroutines if coroutine.cancelled()],
+            )
 
     def join_all_coroutines(self) -> None:
         yield_future(self.wait_for_coroutines())
 
+    def cancel_all_pending(self) -> None:
+        for coroutine in self.coroutines:
+            if not coroutine.done() and not coroutine.cancelled():
+                coroutine.cancel()
+
 
 class RTCPartner(CoroutineHandler):
+    @dataclass
+    class SyncEvents:
+        aio_allow_init: AIOEvent = field(default_factory=AIOEvent)
+        aio_allow_candidates: AIOEvent = field(default_factory=AIOEvent)
+        aio_allow_hangup: AIOEvent = field(default_factory=AIOEvent)
+        reset_event: AIOEvent = field(default_factory=AIOEvent)
```

Comment:
Can you add a bit of documentation how and in which order the events depend on each other?
This feels like it could cause some pain in the future if the events ever get out of sync...